Sources/ServiceDiscovery/ServiceDiscovery.swift

```diff
@@ -12,179 +12,37 @@
+    /// - Returns a ``ServiceDiscoverySubscription`` which produces an ``AsyncSequence`` of changes in  the service discovery instances.
+    /// - throws when failing to establish subscription
+    func subscribe() async throws -> Subscription
 }
 
-// MARK: - Service discovery errors
-
-/// Errors that might occur during lookup.
-public struct LookupError: Error, Equatable, CustomStringConvertible {
-    internal enum ErrorType: Equatable, CustomStringConvertible {
-        case unknownService
-        case timedOut
-
-        var description: String {
-            switch self {
-            case .unknownService:
-                return "unknownService"
-            case .timedOut:
-                return "timedOut"
-            }
-        }
-    }
-
-    internal let type: ErrorType
-
-    public var description: String {
-        "LookupError.\(String(describing: self.type))"
-    }
-
-    /// Lookup cannot be completed because the service is unknown.
-    public static let unknownService = LookupError(type: .unknownService)
-
-    /// Lookup has taken longer than allowed and thus has timed out.
-    public static let timedOut = LookupError(type: .timedOut)
-}
-
-/// General service discovery errors.
-public struct ServiceDiscoveryError: Error, Equatable, CustomStringConvertible {
-    internal enum ErrorType: Equatable, CustomStringConvertible {
-        case unavailable
-        case other(String)
-
-        var description: String {
-            switch self {
-            case .unavailable:
-                return "unavailable"
-            case .other(let detail):
-                return "other: \(detail)"
-            }
-        }
-    }
-
-    internal let type: ErrorType
-
-    public var description: String {
-        "ServiceDiscoveryError.\(String(describing: self.type))"
-    }
-
-    /// `ServiceDiscovery` instance is unavailable.
-    public static let unavailable = ServiceDiscoveryError(type: .unavailable)
-
-    public static func other(_ detail: String) -> ServiceDiscoveryError {
-        ServiceDiscoveryError(type: .other(detail))
-    }
+/// The ServiceDiscoverySubscription returns an AsyncSequence of Result type, with either the
+/// instances discovered or an error if a discovery error has occurred.
+/// The client should decide how to best handle errors in this case, e.g. terminate
+/// the subscription or continue and handle the errors, for example by recording or
+/// propagating them.
+public protocol ServiceDiscoverySubscription: Sendable {
+    /// Service discovery instance type
+    associatedtype Instance: Sendable
+    /// AsyncSequence of Service discovery instances
+    associatedtype DiscoverySequence: Sendable, AsyncSequence where DiscoverySequence.Element == Result<[Instance], Error>
+
+    /// -  Returns a ``DiscoverySequence``, which is an ``AsyncSequence`` where each of its items is a snapshot listing of service instances.
+    func next() async -> DiscoverySequence
```

Comment:
This protocol is a bit weird. I assume your intention is much closer to this:

https://github.com/tomerd/swift-service-discovery/pull/2

Let me know what you think :)